### PR TITLE
Tpetra: Fix #4827; make more Tpetra classes follow Rule of Five

### LINKED ISSUE #4827: Tpetra::Details::copyConvert doesn't build for conversions between Kokkos::complex<float> and double
@trilinos/tpetra 

This breaks `Tpetra::deep_copy` between `MultiVector<std::complex<float>, ...>` and `MultiVector<double, ...>`.

### PR DESCRIPTION
**Pull request by mhoemmen**
@trilinos/tpetra 

## Description

1. Fix #4827.
2. Make more Tpetra classes follow the Rule of Five (C++ Core Guidelines C.21).
3. Clean up FEMultiVector code, syntax, and documentation.

## Related Issues

* Closes #4827 
* Related to #4196, #3946, #3574

### CODE REVIEW

**Comment by mhoemmen** [comment]
Looks legit -- will need to fix something for Stokhos.

**Comment by mhoemmen** [comment]
```
.../Trilinos/packages/tpetra/core/src/Tpetra_Details_copyConvert.hpp:247:43: error: comparison between ‘enum Kokkos::View<Sacado::UQ::PCE<Stokhos::DynamicStorage<int, double, Kokkos::Serial> >*, Kokkos::LayoutRight, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >::<unnamed>’ and ‘enum Kokkos::View<const Sacado::UQ::PCE<Stokhos::DynamicStorage<int, double, Kokkos::Serial> >*, Kokkos::LayoutRight, Kokkos::Device<Kokkos::Serial, Kokkos::HostSpace>, Kokkos::MemoryTraits<0> >::<unnamed>’ [-Werror=enum-compare]
       static_assert (OutputViewType::Rank == InputViewType::Rank,
                      ~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~
```
@etphipp It's interesting that `Rank` is an `int` in all the other `Kokkos::View` specializations.  Should the Sacado etc. specializations behave the same way?

**Comment by mhoemmen** [comment]
Hm, this is interesting: https://testing.sandia.gov/cdash/testDetails.php?test=74057758&build=4851178 .  I can `Kokkos::deep_copy` from LayoutLeft to LayoutRight on the same memory space, but not across memory spaces:
```
 OutputDeviceType: HostSpace
 InputDeviceType: CudaSpace
  OutputArrayLayout: Left
  InputArrayLayout: Left
   OutputValueType: int
   InputValueType: int
   OutputValueType: long long
   InputValueType: int
   OutputValueType: int
   InputValueType: long long
   OutputValueType: Kokkos::complex<float>
   InputValueType: double
   OutputValueType: double
   InputValueType: Kokkos::complex<float>
  OutputArrayLayout: Right
  InputArrayLayout: Left
   OutputValueType: int
   InputValueType: int

p=0: *** Caught standard std::exception of type 'std::runtime_error' :
 
  Error: Kokkos::deep_copy with no available copy mechanism: x to y
  Traceback functionality not available
```

**Comment by etphipp** [comment]
> @etphipp It's interesting that `Rank` is an `int` in all the other `Kokkos::View` specializations. Should the Sacado etc. specializations behave the same way? I don't think this will be hard for me to work around in Tpetra, so there's no urgency.

I'm curious why you say this. From [here](https://github.com/trilinos/Trilinos/blob/ce172471817489c5eb1bdddce8a1948887a6c03f/packages/kokkos/core/src/Kokkos_View.hpp#L610), `Kokkos::View::Rank` is always an enum:
```
  enum { Rank = map_type::Rank };
```
Furthermore, Sacado/Stokhos don't actually specialize `Kokkos::View`, but instead an internal class called `Kokkos::ViewMapping` (which becomes `map_type` in the above).

**Comment by mhoemmen** [comment]
@trilinos/tpetra this should be ready for review now.

File: packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp [hunk 1]
```diff
@@ -1048,12 +1049,58 @@ namespace Tpetra {
...
+    /// \brief Destructor (virtual for memory safety of derived classes).
+    ///
+    /// \note To Tpetra developers: See the C++ Core Guidelines C.21
+    ///   ("If you define or <tt>=delete</tt> any default operation,
+    ///   define or <tt>=delete</tt> them all"), in particular the
+    ///   AbstractBase example, for why this destructor declaration
+    ///   implies that we need the above four <tt>=delete</tt> (or
+    ///   <tt>=default</tt>) declarations for copy construction, move
+    ///   construction, copy assignment, and move assignment.
+    virtual ~CrsMatrix () = default;
```
Comment (lucbv) [inline packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp]: It does seem odd to have the default destructor and not the default constructor... but I understand the historical argument.

Comment (mhoemmen) [reply packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp]: @lucbv https://en.cppreference.com/w/cpp/language/default_constructor#Implicitly-defined_default_constructor :
> If the implicitly-declared default constructor is not defined as deleted, it is defined (that is, a function body is generated and compiled) by the compiler if odr-used, and it has exactly the same effect as a user-defined constructor with empty body and empty initializer list. That is, it calls the default constructors of the bases and of the non-static members of this class.

Comment (mhoemmen) [reply packages/tpetra/core/src/Tpetra_CrsMatrix_decl.hpp]: @lucbv The issue with CrsMatrix is that the copy constructor and copy assignment operator were always declared private and not implemented.  This is equivalent to C++11 `=delete`.  I have no idea what motivated random pre-2010 Tpetra developers to declare them that way, but I feel like fixing that should be a separate PR ;-) .

**Comment by mhoemmen** [comment]
@trilinos/tpetra The above is interesting, because it suggests two things:

1. We weren't using `Kokkos::deep_copy` before, when we could have.
2. We might have been doing the wrong thing before.

@hkthorn @trilinos/anasazi this could actually be a bug.

**Comment by csiefer2** [comment]
"Rule of Five" sounds like something out of Tolkien...

**Comment by mhoemmen** [comment]
> "Rule of Five" sounds like something out of Tolkien...

Best is Rule of Zero i mean they didn't even read a scroll of Detect Curse first

**Comment by mhoemmen** [comment]
I'm working on https://github.com/trilinos/Trilinos/pull/4871 today, but I also want to fix this.

**Comment by mhoemmen** [comment]
@lucbv Thanks for your review!  This PR is superseded by https://github.com/trilinos/Trilinos/pull/4886.  Could you please review that one?  Thanks! :-D

**Pull request closed without merging**